File: go/cms.go
package applereceipt

import (
	"bytes"
	"encoding/asn1"
	"errors"

	"github.com/emindeniz99/apple-purchase-receipt-verifier/go/internal/der"
)

// CMS/PKCS#7 SignedData structure walking for legacy app receipts: the
// part of receipt verification that is pure DER work. Bytes in, bytes
// out — the crypto lives in receipt.go.

var (
	oidSignedData    = asn1.ObjectIdentifier{1, 2, 840, 113549, 1, 7, 2}
	oidMessageDigest = asn1.ObjectIdentifier{1, 2, 840, 113549, 1, 9, 4}
	oidSHA1          = asn1.ObjectIdentifier{1, 3, 14, 3, 2, 26}
	oidSHA256        = asn1.ObjectIdentifier{2, 16, 840, 1, 101, 3, 4, 2, 1}
)

type cmsSignerInfo struct {
	issuerRaw      []byte
	serialContents []byte
	digestOID      asn1.ObjectIdentifier
	signedAttrs    *der.Node // nil when the signature is over the content directly
	signature      []byte
}

type parsedCMS struct {
	content      []byte
	certificates [][]byte
	signerInfo   cmsSignerInfo
}

var errMalformedCMS = errors.New("malformed CMS structure")

// parseCMS walks ContentInfo -> SignedData and pulls out the encapsulated
// content, the embedded certificates and the first SignerInfo.
//
// der.Parse rejects trailing bytes after the outermost value, so a
// receipt with an unverified tail appended is refused here rather than
// silently truncated (PLAN.md §2.3).
func parseCMS(b []byte) (*parsedCMS, error) {
	contentInfo, err := der.Parse(b)
	if err != nil {
		return nil, err
	}
	if contentInfo.Tag != der.TagSequence {
		return nil, errMalformedCMS
	}
	contentType := der.Child(contentInfo, 0)
	wrapper := der.Child(contentInfo, 1)
	if contentType == nil || contentType.Tag != der.TagOID ||
		!oidEqual(contentType.Contents, oidSignedData) ||
		wrapper == nil || wrapper.Tag != der.TagContext0 {
		return nil, errors.New("not a CMS SignedData")
	}
	signedData := der.Child(wrapper, 0)
	if signedData == nil || signedData.Tag != der.TagSequence || len(signedData.Children) < 4 {
		return nil, errMalformedCMS
	}

	// encapContentInfo ::= SEQUENCE { eContentType OID, eContent [0] }
	encap := der.Child(signedData, 2)
	eContent := der.Child(encap, 1)
	if encap == nil || encap.Tag != der.TagSequence || eContent == nil ||
		eContent.Tag != der.TagContext0 {
		return nil, errors.New("no encapsulated payload")
	}
	contentNode := der.Child(eContent, 0)
	if !der.IsOctetString(contentNode) {
		return nil, errors.New("encapsulated payload is not an OCTET STRING")
	}
	content := der.OctetValue(contentNode)

	// certificates [0] IMPLICIT and crls [1] IMPLICIT sit between
	// encapContentInfo and signerInfos, both optional.
	var certificates [][]byte
	fields := signedData.Children
	for _, child := range fields[3 : len(fields)-1] {
		if child.Tag == der.TagContext0 {
			certificates = make([][]byte, 0, len(child.Children))
			for _, cert := range child.Children {
				certificates = append(certificates, cert.Raw)
			}
		}
	}

	signerInfos := fields[len(fields)-1]
	if signerInfos.Tag != der.TagSet || len(signerInfos.Children) == 0 {
		return nil, errors.New("no signer info")
	}
	signerInfo, err := parseSignerInfo(signerInfos.Children[0])
	if err != nil {
		return nil, err
	}
	return &parsedCMS{content: content, certificates: certificates, signerInfo: signerInfo}, nil
}

func parseSignerInfo(node *der.Node) (cmsSignerInfo, error) {
	var info cmsSignerInfo
	if node == nil || node.Tag != der.TagSequence || len(node.Children) < 5 {
		return info, errors.New("malformed SignerInfo")
	}
	sid := der.Child(node, 1)
	issuer := der.Child(sid, 0)
	serial := der.Child(sid, 1)
	if sid == nil || sid.Tag != der.TagSequence || issuer == nil || serial == nil ||
		serial.Tag != der.TagInteger {
		return info, errors.New("SignerInfo does not use issuerAndSerialNumber")
	}
	digestAlgorithm := der.Child(node, 2)
	digestOID := der.Child(digestAlgorithm, 0)
	if digestAlgorithm == nil || digestAlgorithm.Tag != der.TagSequence ||
		digestOID == nil || digestOID.Tag != der.TagOID {
		return info, errors.New("malformed digestAlgorithm")
	}
	oid, err := decodeOID(digestOID.Contents)
	if err != nil {
		return info, err
	}

	index := 3
	var signedAttrs *der.Node
	if attrs := der.Child(node, index); attrs != nil && attrs.Tag == der.TagContext0 {
		signedAttrs = attrs
		index++
	}
	index++ // signatureAlgorithm: the digest algorithm drives the hash
	signature := der.Child(node, index)
	if !der.IsOctetString(signature) {
		return info, errors.New("malformed SignerInfo signature")
	}
	return cmsSignerInfo{
		issuerRaw:      issuer.Raw,
		serialContents: serial.Contents,
		digestOID:      oid,
		signedAttrs:    signedAttrs,
		signature:      der.OctetValue(signature),
	}, nil
}

// findMessageDigestAttribute returns the messageDigest signed attribute's
// value, or nil when the attribute is absent. A signed attribute of the
// wrong shape is an error rather than "not the one we wanted".
func findMessageDigestAttribute(signedAttrs *der.Node) ([]byte, error) {
	for _, attr := range signedAttrs.Children {
		attrType := der.Child(attr, 0)
		values := der.Child(attr, 1)
		value := der.Child(values, 0)
		if attr.Tag != der.TagSequence || attrType == nil || attrType.Tag != der.TagOID ||
			values == nil || value == nil {
			return nil, errors.New("malformed signed attribute")
		}
		if oidEqual(attrType.Contents, oidMessageDigest) {
			return der.OctetValue(value), nil
		}
	}
	return nil, nil
}

// signedAttrsSignedBytes is what the SignerInfo signature actually covers
// when signed attributes are present: the attributes re-encoded as an
// explicit SET (RFC 5652 §5.4), i.e. the IMPLICIT [0] identifier octet
// swapped for SET.
//
// Only a definite-length encoding can be re-tagged this way; a BER
// indefinite-length signedAttrs would need a full re-encode, and no Apple
// receipt uses one.
func signedAttrsSignedBytes(signedAttrs *der.Node) ([]byte, error) {
	raw := signedAttrs.Raw
	if len(raw) < 2 {
		return nil, errors.New("malformed signed attributes")
	}
	if raw[1] == 0x80 {
		return nil, errors.New("indefinite-length signed attributes are not supported")
	}
	out := make([]byte, len(raw))
	copy(out, raw)
	out[0] = der.TagSet
	return out, nil
}

func oidEqual(contents []byte, oid asn1.ObjectIdentifier) bool {
	encoded, err := encodeOIDContents(oid)
	if err != nil {
		return false
	}
	return bytes.Equal(contents, encoded)
}

// encodeOIDContents renders an OID's contents octets, so a comparison can
// be made on bytes rather than by decoding attacker-supplied ones.
func encodeOIDContents(oid asn1.ObjectIdentifier) ([]byte, error) {
	if len(oid) < 2 {
		return nil, errors.New("object identifier too short")
	}
	out := []byte{byte(oid[0]*40 + oid[1])}
	for _, part := range oid[2:] {
		if part < 0 {
			return nil, errors.New("negative object identifier arc")
		}
		var chunk []byte
		value := part
		chunk = append(chunk, byte(value&0x7f))
		value >>= 7
		for value > 0 {
			chunk = append(chunk, byte(value&0x7f)|0x80)
			value >>= 7
		}
		for i := len(chunk) - 1; i >= 0; i-- {
			out = append(out, chunk[i])
		}
	}
	return out, nil
}

// decodeOID reads OID contents octets into an ObjectIdentifier, bounded
// so a hostile arc cannot allocate or spin.
func decodeOID(contents []byte) (asn1.ObjectIdentifier, error) {
	if len(contents) == 0 || len(contents) > 64 {
		return nil, errors.New("unsupported object identifier length")
	}
	first := int(contents[0])
	arc0 := first / 40
	if arc0 > 2 {
		arc0 = 2
	}
	out := asn1.ObjectIdentifier{arc0, first - arc0*40}
	value := 0
	started := false
	for _, b := range contents[1:] {
		if value > (1 << 24) {
			return nil, errors.New("object identifier arc out of range")
		}
		value = value<<7 | int(b&0x7f)
		started = true
		if b&0x80 == 0 {
			out = append(out, value)
			value = 0
			started = false
		}
	}
	if started {
		return nil, errors.New("truncated object identifier")
	}
	return out, nil
}

File: php/src/Internal/Cms.php
<?php

declare(strict_types=1);

namespace EminDeniz99\ApplePurchaseReceiptVerifier\Internal;

use EminDeniz99\ApplePurchaseReceiptVerifier\Reason;
use EminDeniz99\ApplePurchaseReceiptVerifier\VerificationException;

/**
 * CMS/PKCS#7 SignedData structure walking for legacy app receipts — the part
 * of receipt verification that is pure DER work. Bytes in, bytes out: the
 * crypto lives in the verifier.
 *
 * Ported check-for-check from `node/src/cms.ts` so the two implementations
 * cannot disagree about what a receipt contains.
 *
 * @internal
 */
final class Cms
{
    private const OID_SIGNED_DATA = '1.2.840.113549.1.7.2';
    private const OID_MESSAGE_DIGEST = '1.2.840.113549.1.9.4';

    /**
     * Only the digests Apple uses for receipts. Anything else is rejected
     * rather than looked up: a receipt is not the place to be liberal.
     */
    private const DIGEST_ALGORITHMS = [
        '1.3.14.3.2.26' => 'sha1',
        '2.16.840.1.101.3.4.2.1' => 'sha256',
    ];

    /**
     * @param list<string> $certificates DER bytes of each embedded certificate
     * @param 'sha1'|'sha256' $digest the digest the signature is over
     */
    private function __construct(
        public readonly string $content,
        public readonly array $certificates,
        public readonly string $signerIssuerRaw,
        public readonly string $signerSerial,
        public readonly string $digest,
        public readonly ?Asn1Node $signedAttrs,
        public readonly string $signature,
    ) {
    }

    /** @throws VerificationException */
    public static function parse(string $der, int $nodeBudget = Der::DEFAULT_NODE_BUDGET): self
    {
        try {
            $contentInfo = Der::parse($der, $nodeBudget);
        } catch (ParseException $e) {
            throw new VerificationException(Reason::InvalidReceiptFormat, 'not parseable ASN.1', $e);
        }

        try {
            $info = $contentInfo->children();
            if ($contentInfo->tag !== Der::TAG_SEQUENCE
                || count($info) < 2
                || $info[0]->contents !== Der::encodeOidContents(self::OID_SIGNED_DATA)
                || $info[1]->tag !== Der::TAG_CONTEXT_0) {
                throw new ParseException('not a CMS SignedData');
            }
            $inner = $info[1]->child(0);
            if ($inner === null) {
                throw new ParseException('empty SignedData');
            }
            $signedData = $inner->children();
            if (count($signedData) < 4) {
                throw new ParseException('unexpected SignedData layout');
            }
            $encap = $signedData[2]->children();
            if (count($encap) < 2 || $encap[1]->tag !== Der::TAG_CONTEXT_0) {
                throw new ParseException('no encapsulated payload');
            }
            $contentNode = $encap[1]->child(0);
            if ($contentNode === null || !Der::isOctetString($contentNode)) {
                throw new ParseException('encapsulated payload is not an OCTET STRING');
            }
            $content = Der::octets($contentNode);

            $certificates = [];
            $last = count($signedData) - 1;
            for ($i = 3; $i < $last; ++$i) {
                if ($signedData[$i]->tag === Der::TAG_CONTEXT_0) {
                    $certificates = array_map(
                        static fn (Asn1Node $c): string => $c->raw,
                        $signedData[$i]->children(),
                    );
                }
            }

            $signerInfos = $signedData[$last];
            $signerInfo = $signerInfos->child(0);
            if ($signerInfos->tag !== Der::TAG_SET || $signerInfo === null) {
                throw new ParseException('no signer info');
            }

            return self::parseSignerInfo($signerInfo, $content, $certificates);
        } catch (VerificationException $e) {
            throw $e;
        } catch (ParseException $e) {
            throw new VerificationException(Reason::InvalidReceiptFormat, 'malformed CMS structure', $e);
        }
    }

    /**
     * @param list<string> $certificates
     *
     * @throws ParseException
     */
    private static function parseSignerInfo(Asn1Node $node, string $content, array $certificates): self
    {
        $fields = $node->children();
        if (count($fields) < 5) {
            throw new ParseException('unexpected SignerInfo layout');
        }
        $sid = $fields[1]->children();
        if (count($sid) < 2) {
            throw new ParseException('unexpected SignerIdentifier layout');
        }
        $issuerRaw = $sid[0]->raw;
        $serial = $sid[1]->contents;

        $digestOidNode = $fields[2]->child(0);
        if ($digestOidNode === null) {
            throw new ParseException('missing digest algorithm');
        }
        $digestOid = Der::decodeOid($digestOidNode->contents);

        $index = 3;
        $signedAttrs = null;
        if ($fields[$index]->tag === Der::TAG_CONTEXT_0) {
            $signedAttrs = $fields[$index];
            ++$index;
        }
        ++$index; // signatureAlgorithm — RSA PKCS#1 v1.5; the digest OID drives the hash
        if (!isset($fields[$index])) {
            throw new ParseException('unexpected SignerInfo layout');
        }
        $signature = $fields[$index]->contents;

        // Rejected here rather than at signature time so an unsupported digest
        // reads as a malformed receipt, which is what every other port says.
        if (!isset(self::DIGEST_ALGORITHMS[$digestOid])) {
            throw new ParseException('unsupported digest algorithm');
        }

        return new self(
            content: $content,
            certificates: $certificates,
            signerIssuerRaw: $issuerRaw,
            signerSerial: $serial,
            digest: self::DIGEST_ALGORITHMS[$digestOid],
            signedAttrs: $signedAttrs,
            signature: $signature,
        );
    }

    /**
     * Index of the embedded certificate the SignerInfo names by
     * issuerAndSerialNumber, or -1. Unparseable entries are skipped rather
     * than fatal: a receipt may legitimately carry a certificate we cannot
     * read alongside the one we need.
     *
     * @param list<Certificate> $embedded
     */
    public function findSignerIndex(array $embedded): int
    {
        foreach ($embedded as $i => $cert) {
            if (hash_equals($cert->serialNumber, $this->signerSerial)
                && hash_equals($cert->issuerDer, $this->signerIssuerRaw)) {
                return $i;
            }
        }

        return -1;
    }

    /** @throws ParseException */
    public function messageDigestAttribute(): ?string
    {
        if ($this->signedAttrs === null) {
            return null;
        }
        $wanted = Der::encodeOidContents(self::OID_MESSAGE_DIGEST);
        foreach ($this->signedAttrs->children() as $attr) {
            // Every signed attribute is SEQUENCE { OID, SET OF value }; a shape
            // missing either part is malformed, not merely uninteresting.
            $parts = $attr->children();
            $type = $parts[0] ?? null;
            $value = ($parts[1] ?? null)?->child(0);
            if ($type === null || $value === null) {
                throw new ParseException('malformed signed attribute');
            }
            if ($type->contents === $wanted) {
                return $value->contents;
            }
        }

        return null;
    }

    /**
     * The bytes a SignerInfo signature covers when signedAttrs are present:
     * the attributes re-encoded as an explicit SET (RFC 5652 §5.4) — swap the
     * IMPLICIT [0] tag for SET. Signing the `[0]`-tagged bytes as they appear
     * on the wire is the classic mistake here, and it verifies nothing.
     */
    public function signedAttrsSignedBytes(): string
    {
        if ($this->signedAttrs === null) {
            throw new ParseException('no signed attributes');
        }

        return chr(Der::TAG_SET) . substr($this->signedAttrs->raw, 1);
    }
}
